### PR TITLE
[LegalizeVectorOps][RISCV] Don't promote VP_FABS/FNEG/FCOPYSIGN.

### PR DESCRIPTION
**Pull request by topperc**
Promoting canonicalizes NaNs which changes the semantics. Bitcast to integer and use logic ops instead.

### CODE REVIEW

**Review by jacquesguan — APPROVED** [review]
LGTM

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp [hunk 1]
```diff
@@ -1557,6 +1583,80 @@ SDValue VectorLegalizer::ExpandVP_REM(SDNode *Node) {
...
+
+  SDValue Mask = Node->getOperand(1);
+  SDValue EVL = Node->getOperand(2);
+
+  SDLoc DL(Node);
+  SDValue Cast = DAG.getNode(ISD::BITCAST, DL, IntVT, Node->getOperand(0));
+  SDValue SignMask = DAG.getConstant(
+      APInt::getSignMask(IntVT.getScalarSizeInBits()), DL, IntVT);
+  SDValue Xor = DAG.getNode(ISD::VP_XOR, DL, IntVT, Cast, SignMask, Mask, EVL);
+  return DAG.getNode(ISD::BITCAST, DL, VT, Xor);
```
Comment (lukel97) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1601]: How come this expansion is different from VectorLegalizer::ExpandFNEG?

Comment (lukel97) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1601]: I presume because f16 fsub is illegal for zvfhmin?

Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1601]: I'm not sure what version of ExpandFNEG you're looking at. I rewrote it today. aa91d90cb07d72b32176a966fe798ab71ecb0a76

Comment (lukel97) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:1601]: Ah yup, I needed to do a git pull

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp [hunk 2]
```diff
@@ -699,6 +702,11 @@ void VectorLegalizer::Promote(SDNode *Node, SmallVectorImpl<SDValue> &Results) {
     // These operations are used to do promotion so they can't be promoted
     // themselves.
     llvm_unreachable("Don't know how to promote this operation!");
+  case ISD::VP_FABS:
+  case ISD::VP_FCOPYSIGN:
+  case ISD::VP_FNEG:
+    // Promoting fabs, fneg, and fcopysign changes their semantics.
+    llvm_unreachable("These operations should not be promoted");
```
Comment (lukel97) [inline llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:709]: For #106652 do we also need to add an llvm_unreachable for the non VP ops?

**Review by lukel97 — APPROVED** [review]

File: llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp (hunk 2 above)
Comment (topperc) [reply llvm/lib/CodeGen/SelectionDAG/LegalizeVectorOps.cpp:709]: Probably. I haven't checked the other targets yet.

**Comment by preames** [comment]
Purely a side comment... If we knew we had NoNans from fastmath, we could promote.  However, we don't have that information in the opcode, and the resulting codegen from expansion seems better anyways.

**Pull request merged**